cpp/src/gandiva/precompiled/time.cc

```diff
@@ -634,7 +642,7 @@ timestamp castTIMESTAMP_utf8(int64_t context, const char* input, int32 length) {
   }
 
   // adjust the year
-  if (ts_fields[TimeFields::kYear] < 100) {
+  if (ts_fields[TimeFields::kYear] < 100 && year_str_len < 4) {
```

Comment:
Shouldn't one set an error if year length is neither 2 nor 4?

Reply:
year length can vary from 1 - 4. Added test cases for that too.